**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: Yifan Li

## Problem 2:  Practical Implementations of Optimization Functions (27 Points)

In [1]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
% matplotlib inline

C:\Anaconda3\lib\site-packages\h5py\__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


In [2]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

# Create the MLP model
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10]) #true labels

n_hidden_1 = 256 # number of hidden nodes in hidden layer 1
n_hidden_2 = 64 # number of hidden nodes in hidden layer 2
W1 = tf.Variable(initial_value=tf.truncated_normal([784, n_hidden_1], stddev=0.1)) #layer 1
b1 = tf.Variable(tf.zeros([n_hidden_1]))
z1 = tf.nn.relu(tf.matmul(x, W1) + b1)
W2 = tf.Variable(initial_value=tf.truncated_normal([n_hidden_1, n_hidden_2], stddev=0.1)) #layer 2
b2 = tf.Variable(tf.zeros([n_hidden_2]))
z2 = tf.nn.relu(tf.matmul(z1, W2) + b2)
W3 = tf.Variable(initial_value=tf.truncated_normal([n_hidden_2, 10], stddev=0.1)) #layer 3
b3 = tf.Variable(tf.zeros([10]))
y = tf.matmul(z2, W3) + b3

Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.
Instructions for updating:
Please write your own downloading logic.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting MNIST_data/train-images-idx3-ubyte.gz
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Instructions for updating:
Please use tf.one_hot on tensors.
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz
Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.


### (a) SGD

In [3]:
# (a)
# SGD trial 1, stepsize = 1
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(1).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 1 --- stepsize = 1:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 1 --- stepsize = 1:

Iter: 0: validation accuracy: 0.16040000319480896
Iter: 1000: validation accuracy: 0.9714000225067139
Iter: 2000: validation accuracy: 0.9688000082969666
Iter: 3000: validation accuracy: 0.9746000170707703
Iter: 4000: validation accuracy: 0.9783999919891357
Iter: 5000: validation accuracy: 0.9801999926567078
Iter: 6000: validation accuracy: 0.9801999926567078
Iter: 7000: validation accuracy: 0.9796000123023987
Iter: 8000: validation accuracy: 0.9797999858856201
Iter: 9000: validation accuracy: 0.9796000123023987
Iter: 9999: validation accuracy: 0.9787999987602234


In [5]:
# SGD trial 1, stepsize = 0.7
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(0.7).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 2 --- stepsize = 0.7:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys}) # fill in the placeholders
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 2 --- stepsize = 0.7:

Iter: 0: validation accuracy: 0.28279998898506165
Iter: 1000: validation accuracy: 0.9649999737739563
Iter: 2000: validation accuracy: 0.9753999710083008
Iter: 3000: validation accuracy: 0.9764000177383423
Iter: 4000: validation accuracy: 0.978600025177002
Iter: 5000: validation accuracy: 0.9796000123023987
Iter: 6000: validation accuracy: 0.9824000000953674
Iter: 7000: validation accuracy: 0.9775999784469604
Iter: 8000: validation accuracy: 0.9805999994277954
Iter: 9000: validation accuracy: 0.9833999872207642
Iter: 9999: validation accuracy: 0.9828000068664551


In [6]:
# SGD trial 3, stepsize = 0.5
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 3 --- stepsize = 0.5:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys}) # fill in the placeholders
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 3 --- stepsize = 0.5:

Iter: 0: validation accuracy: 0.24420000612735748
Iter: 1000: validation accuracy: 0.9667999744415283
Iter: 2000: validation accuracy: 0.977400004863739
Iter: 3000: validation accuracy: 0.978600025177002
Iter: 4000: validation accuracy: 0.9796000123023987
Iter: 5000: validation accuracy: 0.9842000007629395
Iter: 6000: validation accuracy: 0.9846000075340271
Iter: 7000: validation accuracy: 0.9851999878883362
Iter: 8000: validation accuracy: 0.9843999743461609
Iter: 9000: validation accuracy: 0.9846000075340271
Iter: 9999: validation accuracy: 0.9850000143051147


In [8]:
# SGD trial 4, stepsize = 0.3
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(0.3).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 4 --- stepsize = 0.01:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys}) # fill in the placeholders
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 4 --- stepsize = 0.01:

Iter: 0: validation accuracy: 0.1995999962091446
Iter: 1000: validation accuracy: 0.9688000082969666
Iter: 2000: validation accuracy: 0.9793999791145325
Iter: 3000: validation accuracy: 0.9782000184059143
Iter: 4000: validation accuracy: 0.9805999994277954
Iter: 5000: validation accuracy: 0.9824000000953674
Iter: 6000: validation accuracy: 0.9829999804496765
Iter: 7000: validation accuracy: 0.9836000204086304
Iter: 8000: validation accuracy: 0.984000027179718
Iter: 9000: validation accuracy: 0.9842000007629395
Iter: 9999: validation accuracy: 0.9842000007629395


In [9]:
# SGD trial 5, stepsize = 0.1
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(0.1).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 5 --- stepsize = 0.1:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys}) # fill in the placeholders
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 5 --- stepsize = 0.1:

Iter: 0: validation accuracy: 0.1420000046491623
Iter: 1000: validation accuracy: 0.949999988079071
Iter: 2000: validation accuracy: 0.965399980545044
Iter: 3000: validation accuracy: 0.9706000089645386
Iter: 4000: validation accuracy: 0.9739999771118164
Iter: 5000: validation accuracy: 0.9768000245094299
Iter: 6000: validation accuracy: 0.9775999784469604
Iter: 7000: validation accuracy: 0.9783999919891357
Iter: 8000: validation accuracy: 0.9783999919891357
Iter: 9000: validation accuracy: 0.9793999791145325
Iter: 9999: validation accuracy: 0.9787999987602234


In [10]:
# SGD trial 6, stepsize = 0.01
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.GradientDescentOptimizer(0.01).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD trial 6 --- stepsize = 0.01:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys}) # fill in the placeholders
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD trial 6 --- stepsize = 0.01:

Iter: 0: validation accuracy: 0.09640000015497208
Iter: 1000: validation accuracy: 0.8925999999046326
Iter: 2000: validation accuracy: 0.9156000018119812
Iter: 3000: validation accuracy: 0.9269999861717224
Iter: 4000: validation accuracy: 0.9337999820709229
Iter: 5000: validation accuracy: 0.9401999711990356
Iter: 6000: validation accuracy: 0.9434000253677368
Iter: 7000: validation accuracy: 0.9476000070571899
Iter: 8000: validation accuracy: 0.949999988079071
Iter: 9000: validation accuracy: 0.954800009727478
Iter: 9999: validation accuracy: 0.9563999772071838


### (b) SGD with momentum

In [11]:
# (b)

# SGD with momentum trial 1, stepsize = 0.5, momentum = 0.9
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.MomentumOptimizer(learning_rate=0.5, momentum=0.9).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD w/ momentum trial 1 --- stepsize = 0.5, momentum = 0.9:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD w/ momentum trial 1 --- stepsize = 0.5, momentum = 0.9:

Iter: 0: validation accuracy: 0.22439999878406525
Iter: 1000: validation accuracy: 0.11259999871253967
Iter: 2000: validation accuracy: 0.11259999871253967
Iter: 3000: validation accuracy: 0.11259999871253967
Iter: 4000: validation accuracy: 0.11259999871253967
Iter: 5000: validation accuracy: 0.09860000014305115
Iter: 6000: validation accuracy: 0.10700000077486038
Iter: 7000: validation accuracy: 0.10999999940395355
Iter: 8000: validation accuracy: 0.11259999871253967
Iter: 9000: validation accuracy: 0.0957999974489212
Iter: 9999: validation accuracy: 0.0989999994635582


In [13]:
# SGD with momentum trial 2, stepsize = 0.1, momentum = 0.9
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.MomentumOptimizer(learning_rate=0.1, momentum=0.9).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD w/ momentum trial 2 --- stepsize = 0.1, momentum = 0.9:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD w/ momentum trial 2 --- stepsize = 0.1, momentum = 0.9:

Iter: 0: validation accuracy: 0.07940000295639038
Iter: 1000: validation accuracy: 0.9667999744415283
Iter: 2000: validation accuracy: 0.9742000102996826
Iter: 3000: validation accuracy: 0.9814000129699707
Iter: 4000: validation accuracy: 0.977400004863739
Iter: 5000: validation accuracy: 0.9810000061988831
Iter: 6000: validation accuracy: 0.9807999730110168
Iter: 7000: validation accuracy: 0.9807999730110168
Iter: 8000: validation accuracy: 0.9797999858856201
Iter: 9000: validation accuracy: 0.9746000170707703
Iter: 9999: validation accuracy: 0.9805999994277954


In [14]:
# SGD with momentum trial 3, stepsize = 0.1, momentum = 0.5
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.MomentumOptimizer(learning_rate=0.1, momentum=0.5).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD w/ momentum trial 3 --- stepsize = 0.1, momentum = 0.5:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD w/ momentum trial 3 --- stepsize = 0.1, momentum = 0.5:

Iter: 0: validation accuracy: 0.17399999499320984
Iter: 1000: validation accuracy: 0.9639999866485596
Iter: 2000: validation accuracy: 0.9697999954223633
Iter: 3000: validation accuracy: 0.9769999980926514
Iter: 4000: validation accuracy: 0.979200005531311
Iter: 5000: validation accuracy: 0.9793999791145325
Iter: 6000: validation accuracy: 0.980400025844574
Iter: 7000: validation accuracy: 0.9818000197410583
Iter: 8000: validation accuracy: 0.980400025844574
Iter: 9000: validation accuracy: 0.979200005531311
Iter: 9999: validation accuracy: 0.9832000136375427


In [16]:
# SGD with momentum trial 4, stepsize = 0.05, momentum = 0.9
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.MomentumOptimizer(learning_rate=0.05, momentum=0.9).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD w/ momentum trial 4 --- stepsize = 0.05, momentum = 0.9:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD w/ momentum trial 4 --- stepsize = 0.05, momentum = 0.9:

Iter: 0: validation accuracy: 0.10440000146627426
Iter: 1000: validation accuracy: 0.9648000001907349
Iter: 2000: validation accuracy: 0.972599983215332
Iter: 3000: validation accuracy: 0.9779999852180481
Iter: 4000: validation accuracy: 0.9807999730110168
Iter: 5000: validation accuracy: 0.9796000123023987
Iter: 6000: validation accuracy: 0.9810000061988831
Iter: 7000: validation accuracy: 0.9832000136375427
Iter: 8000: validation accuracy: 0.9847999811172485
Iter: 9000: validation accuracy: 0.9843999743461609
Iter: 9999: validation accuracy: 0.9847999811172485


In [18]:
# SGD with momentum trial 5, stepsize = 0.07, momentum = 0.9
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.MomentumOptimizer(learning_rate=0.07, momentum=0.9).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nSGD w/ momentum trial 5 --- stepsize = 0.07, momentum = 0.9:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


SGD w/ momentum trial 5 --- stepsize = 0.07, momentum = 0.9:

Iter: 0: validation accuracy: 0.09459999948740005
Iter: 1000: validation accuracy: 0.9697999954223633
Iter: 2000: validation accuracy: 0.9765999913215637
Iter: 3000: validation accuracy: 0.9793999791145325
Iter: 4000: validation accuracy: 0.9818000197410583
Iter: 5000: validation accuracy: 0.9801999926567078
Iter: 6000: validation accuracy: 0.9793999791145325
Iter: 7000: validation accuracy: 0.9818000197410583
Iter: 8000: validation accuracy: 0.9801999926567078
Iter: 9000: validation accuracy: 0.9825999736785889
Iter: 9999: validation accuracy: 0.984000027179718


### (c) Adam with default settings

In [19]:
# Adam trial 1, default settings
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer().minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 1 --- default settings:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 1 --- default settings:

Iter: 0: validation accuracy: 0.1324000060558319
Iter: 1000: validation accuracy: 0.9692000150680542
Iter: 2000: validation accuracy: 0.977400004863739
Iter: 3000: validation accuracy: 0.9782000184059143
Iter: 4000: validation accuracy: 0.9760000109672546
Iter: 5000: validation accuracy: 0.9807999730110168
Iter: 6000: validation accuracy: 0.9757999777793884
Iter: 7000: validation accuracy: 0.9814000129699707
Iter: 8000: validation accuracy: 0.9775999784469604
Iter: 9000: validation accuracy: 0.9779999852180481
Iter: 9999: validation accuracy: 0.9797999858856201


### (d) Adam with various stepsizes

In [23]:
# Adam trial 2, stepsize = 0.5
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer(learning_rate=0.5).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 2 --- stepsize = 0.5:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 2 --- stepsize = 0.5:

Iter: 0: validation accuracy: 0.10999999940395355
Iter: 1000: validation accuracy: 0.2134000062942505
Iter: 2000: validation accuracy: 0.11259999871253967
Iter: 3000: validation accuracy: 0.11259999871253967
Iter: 4000: validation accuracy: 0.11259999871253967
Iter: 5000: validation accuracy: 0.0957999974489212
Iter: 6000: validation accuracy: 0.10700000077486038
Iter: 7000: validation accuracy: 0.10019999742507935
Iter: 8000: validation accuracy: 0.09239999949932098
Iter: 9000: validation accuracy: 0.11259999871253967
Iter: 9999: validation accuracy: 0.09860000014305115


In [24]:
# Adam trial 3, stepsize = 0.1
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer(learning_rate=0.1).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 3 --- stepsize = 0.1:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 3 --- stepsize = 0.1:

Iter: 0: validation accuracy: 0.14100000262260437
Iter: 1000: validation accuracy: 0.6850000023841858
Iter: 2000: validation accuracy: 0.5473999977111816
Iter: 3000: validation accuracy: 0.40880000591278076
Iter: 4000: validation accuracy: 0.5590000152587891
Iter: 5000: validation accuracy: 0.4869999885559082
Iter: 6000: validation accuracy: 0.4740000069141388
Iter: 7000: validation accuracy: 0.45260000228881836
Iter: 8000: validation accuracy: 0.37880000472068787
Iter: 9000: validation accuracy: 0.43939998745918274
Iter: 9999: validation accuracy: 0.257999986410141


In [25]:
# Adam trial 4, stepsize = 0.01
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer(learning_rate=0.01).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 4 --- stepsize = 0.01:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 4 --- stepsize = 0.01:

Iter: 0: validation accuracy: 0.21860000491142273
Iter: 1000: validation accuracy: 0.9607999920845032
Iter: 2000: validation accuracy: 0.9684000015258789
Iter: 3000: validation accuracy: 0.9739999771118164
Iter: 4000: validation accuracy: 0.9711999893188477
Iter: 5000: validation accuracy: 0.968999981880188
Iter: 6000: validation accuracy: 0.974399983882904
Iter: 7000: validation accuracy: 0.9721999764442444
Iter: 8000: validation accuracy: 0.9715999960899353
Iter: 9000: validation accuracy: 0.9757999777793884
Iter: 9999: validation accuracy: 0.9729999899864197


In [26]:
# Adam trial 5, stepsize = 0.001
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer(learning_rate=0.001).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 5 --- stepsize = 0.001:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 5 --- stepsize = 0.001:

Iter: 0: validation accuracy: 0.16740000247955322
Iter: 1000: validation accuracy: 0.9715999960899353
Iter: 2000: validation accuracy: 0.9753999710083008
Iter: 3000: validation accuracy: 0.9797999858856201
Iter: 4000: validation accuracy: 0.980400025844574
Iter: 5000: validation accuracy: 0.9807999730110168
Iter: 6000: validation accuracy: 0.9805999994277954
Iter: 7000: validation accuracy: 0.9819999933242798
Iter: 8000: validation accuracy: 0.9782000184059143
Iter: 9000: validation accuracy: 0.9814000129699707
Iter: 9999: validation accuracy: 0.978600025177002


In [27]:
# Adam trial 6, stepsize = 0.0005
# Define loss and optimizer
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y_, logits=y))
train_step = tf.train.AdamOptimizer(learning_rate=0.0005).minimize(cross_entropy)

# Learning
sess = tf.Session()
sess.run(tf.global_variables_initializer())
print('\nAdam trial 6 --- stepsize = 0.0005:\n')

for i in range(10000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
    if i % 1000 == 0 or i == 9999: # Evaluation
        correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        print('Iter: {}: validation accuracy: {}'.format(i, sess.run(accuracy, feed_dict={x: mnist.validation.images, y_: mnist.validation.labels})))


Adam trial 6 --- stepsize = 0.0005:

Iter: 0: validation accuracy: 0.10920000076293945
Iter: 1000: validation accuracy: 0.9617999792098999
Iter: 2000: validation accuracy: 0.9732000231742859
Iter: 3000: validation accuracy: 0.978600025177002
Iter: 4000: validation accuracy: 0.9801999926567078
Iter: 5000: validation accuracy: 0.977400004863739
Iter: 6000: validation accuracy: 0.9793999791145325
Iter: 7000: validation accuracy: 0.9814000129699707
Iter: 8000: validation accuracy: 0.9811999797821045
Iter: 9000: validation accuracy: 0.9824000000953674
Iter: 9999: validation accuracy: 0.9818000197410583


### Short Answers
** (a) **
Qualitatively, all 4 algorithms (normal SGD, SGD with momentum, default Adam, Adam with various stepsizes) have pretty decent outputs. Each one of these can reach an accuracy of ~98%.

Based on the results shown above, the optimal constant stepsize for my multilayer perceptron classifier with 2 hidden layers is about 0.5. Actually, the validation accuracy reaches ~98% after stepsize increases to 0.1, and accuracy continues to increase as learning rate increases. After a learning rate threshold of around 0.5, the accuracy starts to decrease. We can see that the validation accuracy is not fully optimized when stepsize is either too large (1) or too small (0.01).

The optimal settings for the SGD optimizer with momentum for my multilayer perceptron classifier is: learning rate = 0.05, momentum = 0.9. In fact, the accuracy is pretty high (~98%) when learning rate is about 0.1 and momentum is 0.9. The accuracy slightly increases as I continue to lower these two parameters. However, a relatively large learning rate of ~0.5 yields a very bad accuracy of 10%.

The Adam optimizer with default settings could provide an accuracy of ~98%. And Adam optimizer is quite sensitive to learning rates. For instance, a learning rate of 0.1 is common to use on other optimizers that could generate a decent output, but without changing other parameters, the learning rate of 0.1 will generate a horrible accuracy if used on Adam Optimizer. Generally speaking, when testing on MNIST using my MLP, a smaller learning rate is better for Adam optimizer.

** (b) **
It is hard to fine-tune the optimal settings for different optimizers. But as long as a acceptable output is generated, I could follow the general trends when tweaking the parameters. When dealing with diffrent datasets, different maching learning models, and different gradient descent optimizers, it will be difficult to find a good entry point.

** (c) ** 
As mentioned above in part (a), all 4 algorithms could perform very well. All of them could reach an validation accuracy of ~98% for MNIST dataset trained on my MLP with 2 hidden layers. Among all the trials, default Adam and SGD with momentum have better overall accuracy when trying with different learning rates. Adam with various learning rates is very sensitive to learning rate if other parameters are set to default. The normal SGD also performs well, but the learning rate is the highest.



## Problem 3: Bookkeeping (3 points)

### (a) ~ 4 hours

### (b)
**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: Yifan Li

I adhered to the Duke Community Standard in the completion of this assignment.